webapp/src/Service/PointsScoreService.php

```diff
@@ -0,0 +1,96 @@
+<?php declare(strict_types=1);
+
+namespace App\Service;
+
+use App\Entity\Contest;
+use App\Entity\ContestProblem;
+use App\Entity\JudgeTask;
+use App\Entity\Judging;
+use App\Entity\JudgingRun;
+use App\Entity\Language;
+use App\Entity\Problem;
+use App\Entity\Submission;
+use App\Entity\SubmissionFile;
+use App\Entity\Team;
+use App\Entity\User;
+use App\Utils\FreezeData;
+use App\Utils\Utils;
+use Doctrine\DBAL\Exception as DBALException;
+use Doctrine\ORM\EntityManagerInterface;
+use Doctrine\ORM\NonUniqueResultException;
+use Doctrine\ORM\NoResultException;
+use Doctrine\ORM\Query\Expr\Join;
+use InvalidArgumentException;
+use PHPUnit\Exception;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\HttpFoundation\File\UploadedFile;
+use Symfony\Component\HttpFoundation\StreamedResponse;
+use Symfony\Component\HttpKernel\Exception\AccessDeniedHttpException;
+use Symfony\Component\HttpKernel\Exception\BadRequestHttpException;
+use Symfony\Component\HttpKernel\Exception\ServiceUnavailableHttpException;
+use ZipArchive;
+
+/**
+ * Class SubmissionService
+ * @package App\Service
+ */
+class PointsScoreService
+{
+    protected EntityManagerInterface $em;
+    protected LoggerInterface $logger;
+    protected DOMJudgeService $dj;
+    protected ConfigurationService $config;
+    protected EventLogService $eventLogService;
+    protected ScoreboardService $scoreboardService;
+
+    public function __construct(
+        EntityManagerInterface $em,
+        LoggerInterface        $logger,
+        DOMJudgeService        $dj,
+        ConfigurationService   $config,
+        EventLogService        $eventLogService,
+        ScoreboardService      $scoreboardService
+    )
+    {
+        $this->em = $em;
+        $this->logger = $logger;
+        $this->dj = $dj;
+        $this->config = $config;
+        $this->eventLogService = $eventLogService;
+        $this->scoreboardService = $scoreboardService;
+    }
+
+    /**
+     * @param JudgingRun[] $judgingRuns
+     */
+    public static function getScoredPoints(Judging         $judging,
+                                           array           $judgingRuns,
+                                           ContestProblem  $contestProblem,
+                                           DOMJudgeService $dj): float
+    {
+        if ($judging->getResult() === 'correct') {
+            return $contestProblem->getPoints();
+        }
+
+        $pointsScored = 0;
+
+        foreach ($contestProblem->getProblem()->getTestcaseGroups() as $testcaseGroup) {
```

Comment:
aha, tady se ti hodi mit link mezi problemem a pouzitymi skupinami testacase. Chtel bych odstranit to propojeni (jelikoz mi stale prijde zbytecne). Predpokladam, ze se zde hodi mit strukturu neco jako pole, kde klic je skupina a hodnota je pole judging runu, ktere do dane skupiny patri.

Tuto logiku bych se vubec nebal dat tady pred ten foreach a pomoci array_reduce tuto strukturu vytvorit jednim linearnim pruchodem pres judging runy.

Muzes pak iterovat pres tuto strukturu a jako key budes mit treba id te skupiny a jako values budes mit array judging runu. Ve vnitrku tohoto foreache muzres iterovat pres ty prirazene judging runy a pokud vsechny z tech judging runu maji "currect", priradit prislusny pocet bodu.

Aby jsi zde nemusel pouzivat continue 2 (coz je z meho pohledu velmi tezce citelne, muzes zase pouzit array_reduce, kde budes pocitat (pricitat), kolikrat se v iterovanem judgin runu vyskytuje "correct".
pseudokod neco jako

```
$correctRuns = array_reduce(fn ($carry, $run) => $run->getRunresult() == 'correct' ? $carry + 1 : $carry, 0)
```
a potom tuto hodnotu otestovat na rovnost oproti count($struktura[$groupid]) nebo tak neco

php >8.0 ma weakmap struktury, kde muzes pouzivat objekty jako klice do asociativniho pole, ale myslim, ze aktualni DJ ma pouze 7.4, takze budes muset mit jako klic v te strukture idcko te skupiny